services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/SchemaDiffTraverser.java

```diff
@@ -0,0 +1,246 @@
+            targetSchema == null ? null : generateField(fieldName, targetSchema)));
+  }
+
+  private boolean isNestedType(Schema schema) {
+    return EnumSet.of(Schema.Type.RECORD, Schema.Type.UNION, Schema.Type.ARRAY, Schema.Type.MAP)
+        .contains(schema.getType());
+  }
+
+  private static Schema.Field generateField(String name, Schema schema) {
+    return new Schema.Field(name, schema, "", null);
+  }
+
+  private static String buildFieldPath(String parent, String field) {
+    return parent.isEmpty() ? field : parent + "_" + field;
+  }
+
+  public BiFunction<Object, Pair<Schema.Field, Schema.Field>, Boolean> createSemanticCheck(
+      AtomicReference<String> errorMessage) {
+    return (object, schemasPair) -> {
+      Schema.Field currentField = schemasPair.getFirst();
+      Schema.Field targetField = schemasPair.getSecond();
+
+      if (currentField == null || object == null)
+        return false;
+
+      if (targetField == null) {
+        if (isNestedType(currentField.schema())) {
+          errorMessage.set(craftErrorMessage(currentField.name(), object));
+          return true;
+        }
+        if (isNonDefaultValue(object, currentField, errorMessage))
+          return true;
+      } else {
+        if (AvroSchemaUtils.compareSchemaIgnoreFieldOrder(currentField.schema(), targetField.schema()))
+          return false;
+
+        switch (currentField.schema().getType()) {
+          case ENUM:
+            if (isNewEnumValue(object, currentField, targetField, errorMessage))
+              return true;
+            break;
+          case FIXED:
+            break;
+          default:
+            if (isNonDefaultValue(object, currentField, errorMessage))
+              return true;
+            break;
+        }
+      }
+      return false;
+    };
+  }
+
+  private boolean isNewEnumValue(
+      Object object,
+      Schema.Field currentField,
+      Schema.Field targetField,
+      AtomicReference<String> errorMessage) {
+    List<String> currentSymbols = currentField.schema().getEnumSymbols();
+    List<String> targetSymbols = targetField.schema().getEnumSymbols();
+    if (targetSymbols.stream().anyMatch(symbol -> !currentSymbols.contains(symbol) && symbol.equals(object))) {
+      errorMessage.set(craftErrorMessage(currentField.name(), object));
+      return true;
+    }
+    ;
+    return false;
+  }
+
+  private boolean isNonDefaultValue(Object value, Schema.Field field, AtomicReference<String> errorMessage) {
+    Map<Schema.Type, Object> defaultValues = new HashMap<>();
+    defaultValues.put(Schema.Type.STRING, "");
+    defaultValues.put(Schema.Type.BOOLEAN, false);
+    defaultValues.put(Schema.Type.INT, 0);
+    defaultValues.put(Schema.Type.LONG, 0L);
+    defaultValues.put(Schema.Type.FLOAT, 0.0f);
+    defaultValues.put(Schema.Type.DOUBLE, 0.0d);
+    defaultValues.put(Schema.Type.BYTES, new byte[0]);
+    if (value != null && !value.equals(defaultValues.getOrDefault(field.schema().getType(), null))
```

Comment:
can we move the null-check on the value before creating the map?

Reply:
Updated, thank you for your suggestion.